Update to Python=3.9 and update Django

jonwzheng opened the pull request:
This PR seeks to update `RMG-website` to be compatible with Python 3.9 so that it can leverage the newest version of `RMG-Py` and other packages.

To do this, a number of things need to be updated simultaneously (avoid dependency hell):
* Either update `solprop` using the py3.9 compatible version, or turn it into an API service so that it can be queried separately. Updating `solprop` would for now be the most straightforward option. Although an API service would be the best solution technologically, it would take the most technical load to learn and implement sustainably. Another way is to call it via a subprocess in its own separate `conda` or `docker` environment, which would also require some rewriting of the backend. Key concern is whether it's fast enough. In the future, probably we want to move toward one of the two latter options, especially if we want to deploy more ML models on the site.
* Update Django from 2.2 and then see what breaks; then fix those errors in the codebase.
* See if any new package versions break existing code

There's also an opportunity to make the `conda` environment leaner (which should make solving for it more straightforward in general).

edit 23 March 2026 by @JacksonBurns:

@jonwzheng and I have collaboratively made efforts to move `solprop` into a container and use it as a microservice. This allows us to remain on the old Python 3.7 version and appreciably simplify the RMG-website environment.

jonwzheng committed:
greatly loosen environment restrictions

jonwzheng committed:
add back chemprop_solvation

jonwzheng committed:
clarify solvation imports

jonwzheng committed:
replace deprecated ifequal javascript with if

jonwzheng committed:
fix django, solvation code

jonwzheng committed:
update gitignore with removal of apache

jonwzheng committed:
fix faulty byte decode

jonwzheng committed:
remove java from fluxdiagram

jonwzheng committed:
update solubility code with workaround for argparse

jonwzheng committed:
progress toward making solprop work

jonwzheng committed:
pin django to 4.2

jonwzheng committed:
remove load solubility code as no longer needed

jonwzheng committed:
restore graph vis, ensuring conda env is in path

jonwzheng committed:
make migrations

jonwzheng committed:
update django setting for django3

jonwzheng committed:
update deprecated ajax code for Django 3.1

jonwzheng commented:
Unrelated note, but the entryEdit functionality seems kind of suspicious and maybe we shouldn't allow users to actually use those (I don't think anyone is, and they don't even work on the main site.) We may want to just remove it.

jonwzheng committed:
refactor views code

jonwzheng commented:
Some SoluteML functionalities may be present in SolProp and thereby maybe we don't need chemprop_solvation.

jonwzheng committed:
Update environment file

- update descriptastorus old pin (was causing issues/outdated)
- update solprop mix to new repo
- add cantera pin because of weird glitch?

jonwzheng committed:
update env file with wandb req for solprop

jonwzheng committed:
use chemprop_solvation for SoluteML estimator and not solprop.

jonwzheng committed:
Update how solubility data is loaded

jonwzheng committed:
fix: solvent_smiles and solute_smiles names in solub data loading

jonwzheng committed:
Update solubility code to be compatible

Note, reference solubility data is not yet available.
This will require changes to SolProp

jonwzheng committed:
Add tags suggesting users to use fastsolv

Solprop still has some issues so fastsolv might be a good alternative in
the interim. Also we just recommend it because it's a good model.

BonhyeokKoo commented:
I've confirmed that SoluteML is returning unphysical values in the current development version. This is the same issue described in fhvermei/chemprop_solvation#3, caused by changes in newer `descriptastorus` versions.

What issues did keeping the older pin of `descriptastorus` cause?

jonwzheng commented:
Another option could be to not use `chemprop_solvation` at all and just use solprop_ml which I think has a re-trained version of SoluteML: https://anaconda.org/simonabuzzi/solprop_ml_mix

This would require some changes to the code though.

jonwzheng committed:
Pin scipy version and descriptastorus

We observed non-physical SoluteML values due presumably to a new version
of descripastorus.
However, old versions use the old name of "gilbrat" instead of "gibrat",
a name that was removed in scipy 1.11.0. And at the same time RMG
requires scipy >=1.9.0 for MILP functionality.

jonwzheng commented:
Testing other versions of `descriptastorus`, none of them are compatible:
* 2.0.6 and earlier have the same `scipy` problem
* Later versions have some other packaging issues.

I see we can get descripastorus=2.5.0 by pinning the scipy version (which isn't a tenable long-term solution, but the env can at least be solved.)

Unfortunately, it looks like this still results in non-physical answers. Even trying `descriptastorus=2.2.0` results in the non-physicality. So maybe it's something else?

jonwzheng committed:
add note indicating these models were re-trained

jonwzheng committed:
WIP: replace SoluteML with solprop_ml_mix

SoluteML package seems busted. We might need to rely on the newly
retrained version instead.

jonwzheng committed:
working code for single soluteML

jonwzheng committed:
Refactor DirectML solute to work for batch inputs

The new soluteML version code can process a batch of inputs, so the code is
rewritten to accommodate this.

jonwzheng committed:
Fix typo in DirectML

Gsolv model was called instead of Hsolv

jonwzheng committed:
solprop_ml_mix revert

Reverts "WIP: replace SoluteML with solprop_ml_mix"

This reverts commit 063fbdff12f05521fee51143ad8cfdb9d961bc31.

jonwzheng committed:
Use original version of solprop in a new microservice

JacksonBurns committed:
undo example change

jonwzheng committed:
Update SoluteML

After updating to microservices (many thanks to you Jackson!),
there was a remaining mismatch in data types that caused an error.
This commit fixes it, so the user can use SoluteML now.

JacksonBurns reviewed (approved):
From what (little) I know about web dev, this looks good. The proof is in the pudding, and the pudding runs smoothly and more securely than ever before.

Going to talk on one last commit to update a missing link in the microservice documentation, then you can merge this at your pleasure @jonwzheng. Thanks again for sticking around to help push this over the finish line

JacksonBurns committed:
solve evolution (add missing link)
